model/PermissionsService.php

```diff
@@ -54,61 +52,148 @@ public function __construct(
         $this->eventManager = $eventManager;
     }
 
-    public function saveResourcePermissionsRecursive(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet
-    ): void {
-        $this->saveResourcePermissions($resource, $privilegesToSet, true);
-    }
-
-    private function saveResourcePermissions(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet,
-        bool $isRecursive
-    ): void {
-        $currentPrivileges = $this->dataBaseAccess->getResourcePermissions($resource->getUri());
-        $addRemove = $this->strategy->normalizeRequest($currentPrivileges, $privilegesToSet);
+    public function applyPermissions(ChangePermissionsCommand $command): void
+    {
+        $resources = $this->getResourcesToUpdate($command);
+        $currentPermissions = $this->getResourcesPermissions($resources);
+        $permissionsDelta = $this->getDeltaForResourceTree($command, $currentPermissions);
 
-        if (empty($addRemove)) {
+        if (empty($permissionsDelta)) {
             return;
         }
 
-        $resourcesToUpdate = $this->getResourcesToUpdate($resource, $isRecursive);
-        $permissionsList = $this->getResourcesPermissions($resourcesToUpdate);
-        $actions = $this->getActions($resourcesToUpdate, $permissionsList, $addRemove);
-
-        $this->dryRun($actions, $permissionsList);
+        $actions = $this->getActions($resources, $currentPermissions, $permissionsDelta);
+        $this->dryRun($actions, $currentPermissions);
         $this->wetRun($actions);
-        $this->triggerEvents($addRemove, $resource->getUri(), $isRecursive);
+
+        $this->triggerEvents(
+            $permissionsDelta[$command->getRoot()->getUri()],
+            $command->getRoot()->getUri(),
+            $command->isRecursive(),
+            $command->applyToNestedResources()
+        );
+    }
+
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
+    public function saveResourcePermissionsRecursive(
+        core_kernel_classes_Resource $resource,
+        array $privilegesToSet
+    ): void {
+        $command = new ChangePermissionsCommand($resource, $privilegesToSet);
+        $command->withRecursion();
+        $command->withNestedResources();
+
+        $this->applyPermissions($command);
     }
 
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
     public function savePermissions(
         bool $isRecursive,
         core_kernel_classes_Class $class,
         array $privilegesToSet
     ): void {
-        $this->saveResourcePermissions($class, $privilegesToSet, $isRecursive);
+        $command = new ChangePermissionsCommand($class, $privilegesToSet);
+
```

Comment:
suggestion: same here, i tend to use immutable objects, i would check the $isRecursive and create instances with or without recursions calling new on both cases instead of creating and then setting missing attributes.

this approach is caused because for what i see the ChangePermissionsCommand may be in a wrong state until its set up correctly (while i dont execute the  $command->withRecursion() it will remain as false, which doesn't represent the true state of the business)

Reply:
Answered in [the thread about using mutators](https://github.com/oat-sa/extension-tao-dac-simple/pull/211#discussion_r1134003284).

Reply:
I do prefer the constructor too, but since it can be "opiniated", I think we can let as is and move forward. All good? :)